### PR TITLE
LuaVAO Instanced Models

### PR DESCRIPTION
**Pull request by Krogoth100**
+vao:ModifyUnitBins
+vao:ModifyFeatureBins
+vao:SubmitBins

Bins are placed in a vector instead of using map directly, it should have some advantages (For one, vector is transparent)

As of now it is simplified: only InstanceData, no other instance attributes. I will be expanding it to work with arbitrary instance data a bit later.

W.I.P.

### CODE REVIEW

**Comment by Beherith** [comment]
How does this relate to the push-pop of instanceVBOidTable ?

**Comment by lhog** [comment]
A few comments:

- I think it's better to stick to separate Add/Remove functions
- Why did you add explicit (but optional) count of elements?

File: rts/Lua/LuaVAOImpl.cpp [hunk 1]
```diff
@@ -185,6 +194,116 @@ int LuaVAOImpl::AddObjectsToSubmissionImpl(const sol::stack_table& ids)
...
+template<typename TObj>
+void LuaVAOImpl::Bins::ModifyImpl(const sol::stack_table& removedObjects, const sol::stack_table& addedObjects,
+	sol::optional<size_t> removedCount, sol::optional<size_t> addedCount,
+	std::vector<SDrawElementsIndirectCommand>& submitCmds)
+{
+	const size_t removedObjectCount = (removedCount.has_value()? removedCount.value() : removedObjects.size());
+	const size_t addedObjectCount = (addedCount.has_value()? addedCount.value() : addedObjects.size());
+	if (removedObjectCount == 0 && addedObjectCount == 0) return;
+
+	size_t firstChangedBinIndex = -1;
```
Comment (lhog) [inline rts/Lua/LuaVAOImpl.cpp:206]: This tripped me up a bit until I saw it was size_t, better go for `std::numeric_limits<size_t>::max()`

File: rts/Lua/LuaVAOImpl.cpp [hunk 2]
```diff
@@ -185,6 +194,116 @@ int LuaVAOImpl::AddObjectsToSubmissionImpl(const sol::stack_table& ids)
...
+
+		objIdToLocalInstance[objId] = bin.objIds.size();
+		bin.objIds.emplace_back(objId);
+		const auto matIndex = matrixUploader.GetElemOffset(obj);
+		const auto uniIndex = modelsUniformsStorage.GetObjOffset(obj);
+		const size_t bposeIndex = matrixUploader.GetElemOffset(model);
+		bin.instanceData.emplace_back(
+			static_cast<uint32_t>(matIndex),
+			static_cast<uint8_t>(obj->team), // initial team id, will not be updated (unreliable: shader must obtain it somehow else, otherwise it will not reflect transfer to another team)
+			static_cast<uint8_t>(0), // no draw flags, they will not be updated
```
Comment (lhog) [inline rts/Lua/LuaVAOImpl.cpp:276]: Please still submit initial drawflags in order to adhere to the existing conventions.

File: rts/Lua/LuaVAOImpl.cpp [hunk 3]
```diff
@@ -185,6 +194,116 @@ int LuaVAOImpl::AddObjectsToSubmissionImpl(const sol::stack_table& ids)
 	return submitCmds.size() - idsSize;
 }
 
+template<typename TObj>
+void LuaVAOImpl::Bins::ModifyImpl(const sol::stack_table& removedObjects, const sol::stack_table& addedObjects,
+	sol::optional<size_t> removedCount, sol::optional<size_t> addedCount,
+	std::vector<SDrawElementsIndirectCommand>& submitCmds)
```
Comment (lhog) [inline rts/Lua/LuaVAOImpl.cpp:200]: This should belong to the class and accessed directly, no need to carry it along with arguments.

**Comment by lhog** [comment]
The implementation generally looks bulkier than it should, are you sure you can't reuse bits and pieces from https://github.com/beyond-all-reason/spring/blob/BAR105/rts/Rendering/Models/3DModelVAO.cpp#L270-L304 ?

**Comment by Krogoth100** [comment]
> How does this relate to the push-pop of instanceVBOidTable ?

I think, this serves same purpose, except it is fully instanced. Basic functions are not instanced in a sense that they do not instance index data, they push indices every time a repetitive model is pushed.

Pick objects via RenderXDrawFlagChanged > process > send to bins of respective VAOs.

File: rts/Lua/LuaVAOImpl.cpp (hunk 1 above)
Comment (Krogoth100) [reply rts/Lua/LuaVAOImpl.cpp:206]: Ok

**Comment by Krogoth100** [comment]
> The implementation generally looks bulkier than it should, are you sure you can't reuse bits and pieces from https://github.com/beyond-all-reason/spring/blob/BAR105/rts/Rendering/Models/3DModelVAO.cpp#L270-L304 and https://github.com/beyond-all-reason/spring/blob/BAR105/rts/Rendering/Models/3DModelVAO.cpp#L344-L389 ?

First yes, I should probably make instance data external function.
Second has significant differences, hard to reuse.

File: rts/Lua/LuaVAOImpl.cpp (hunk 3 above)
Comment (Krogoth100) [reply rts/Lua/LuaVAOImpl.cpp:200]: > This should belong to the class and accessed directly, no need to carry it along with arguments.

The thing is, submitCmds is the only thing that is used from LuaVAO here. All the other fields are good to work this way.
Maybe move this thing in another cpp altogether?

**Comment by Krogoth100** [comment]
> A few comments:
> 
> * I think it's better to stick to separate Add/Remove functions
> * Why did you add explicit (but optional) count of elements?

Thought of that, that doesn't make practical sense. Usual routine is to collect added and removed objects and just "push changes". No separate Add/Remove make much meaning.

Explicit count because usually Lua knows it, it builds the arrays, no need to call size().

**Comment by Krogoth100** [comment]
@lhog
Addressed some of the issues

+GetObjectInstanceData
a useful one, but I don't know where to put it

I am thinking of separating Bins into another file, not sure how yet, because the code is gonna grow.
Also I want ModifyImpl to be a part of Bins and not of general Vao, as it logically is.

**Comment by Krogoth100** [comment]
@lhog
Planned to do more, but I think that's it for this PR
Only need to sort out GetObjectInstanceData

**Comment by lhog** [comment]
This is better, but Add/Remove is a must otherwise it creates disconnect to the current API, also remove element counts, it's a potential source of errors, that doesn't bring any tangible value.

**Pull request closed without merging**